internal/datastore/mysql/revisions.go

```diff
@@ -0,0 +1,178 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/shopspring/decimal"
+
+	"github.com/authzed/spicedb/internal/datastore"
+)
+
+// querySelectRevision will round the database's timestamp down to the nearest
+// quantization period, and then find the first transaction after that. If there
+// are no transactions newer than the quantization period, it just picks the latest
+// transaction. It will also return the amount of nanoseconds until the next
+// optimized revision would be selected server-side, for use with caching.
+//
+//   %[1] Name of id column
+//   %[2] Relationship tuple transaction table
+//   %[3] Name of timestamp column
+//   %[4] Quantization period (in nanoseconds)
+const querySelectRevision = `SELECT COALESCE((
+  SELECT MIN(%[1]s)
+  FROM   %[2]s
+  WHERE  %[3]s >= FROM_UNIXTIME(FLOOR(UNIX_TIMESTAMP(UTC_TIMESTAMP(6)) * 1000000000 / %[4]d) * %[4]d / 1000000000)
+), (
+  SELECT MAX(%[1]s)
+  FROM   %[2]s
+)) as revision,
+%[4]d - CAST(UNIX_TIMESTAMP(UTC_TIMESTAMP(6)) * 1000000000 AS UNSIGNED INTEGER) %% %[4]d as validForNanos;`
+
+// queryValidTransaction will return a single row with two values, one boolean
+// for whether the specified transaction ID is newer than the garbage collection
+// window, and one boolean for whether the transaction ID represents a transaction
+// that will occur in the future.
+//
+//   %[1] Name of id column
+//   %[2] Relationship tuple transaction table
+//   %[3] Name of timestamp column
+//   %[4] Inverse of GC window (in seconds)
+const queryValidTransaction = `SELECT ? >= (
+  SELECT MIN(%[1]s)
+  FROM   %[2]s
+  WHERE  %[3]s >= TIMESTAMPADD(SECOND, %.6[4]f, UTC_TIMESTAMP(6))
+) as fresh, ? > (
+  SELECT MAX(%[1]s)
+  FROM   %[2]s
+) as future;`
+
+func (mds *Datastore) optimizedRevisionFunc(ctx context.Context) (datastore.Revision, time.Duration, error) {
+	var revision uint64
+	var validForNanos time.Duration
+	if err := mds.db.QueryRowContext(
+		datastore.SeparateContextWithTracing(ctx), mds.optimizedRevisionQuery,
+	).Scan(&revision, &validForNanos); err != nil {
+		return datastore.NoRevision, 0, fmt.Errorf(errRevision, err)
+	}
+	return revisionFromTransaction(revision), validForNanos, nil
+}
+
+func (mds *Datastore) HeadRevision(ctx context.Context) (datastore.Revision, error) {
+	// implementation deviates slightly from PSQL implementation in order to support
+	// database seeding in runtime, instead of through migrate command
+	ctx, span := tracer.Start(ctx, "HeadRevision")
+	defer span.End()
+
+	revision, err := mds.loadRevision(ctx)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+	if revision == 0 {
+		return datastore.NoRevision, nil
```

Comment:
Why return `NoRevision` here vs. just returning the `0` revision?

Reply:
`loadRevision()` can return a revision of `0` with no error indicating the database hasn't yet been seeded.

https://github.com/authzed/spicedb/blob/b33fe794326461cd95cc77f3c106974818e83946/internal/datastore/mysql/datastore.go#L643-L654

Returning `0` here would imply the database was seeded with the initial value of `0`, which would be incorrect. Instead, we return `NoRevision` and no error to signal there isn't a revision but also no error.

Reply:
I haven't looked at the seeding part much, would it make sense to return an error in that case?

Anyway that's not really an issue with this PR so I'll resolve this.

Reply:
Yeah, that probably would make sense. The `isSeeded` function is the one checking for no error and `NoRevision` to trigger seeding

https://github.com/authzed/spicedb/blob/a86dccf77489314d6cc3ec1fe1f722330865ff06/internal/datastore/mysql/datastore.go#L454-L470

If we refactor this to return an error, `isSeeded` will have to inspect the error and check if it's equal to `ErrNoRevisionSeed` or some such.